features/5-basic-sso/api.md

```diff
@@ -0,0 +1,207 @@
+- `-[SKYContainer associateAccountWithProvider:(NSString*)providerID options:(NSDictionary*)options completion:(void(^)(NSError*, SKYUser*))]`
+  - Add a new auth provider to the user by going through the auth flow
+  - `providerID` - A string that identify the login provider
+  - `options`
+    - uxMode - Either `popup`(default), or `redirect`, popup means in-app-browser (SFSafariViewController/WKWebView) and redirect means Safari.app
+    - clientID
+    - scope
+    - version - FB Client SDK only
+    - cookiePolicy - Google Client SDK only
+  - This function returns a skygear user, and an access token of the new service, via delegate.
+- `-[SKYContainer getOAuthTokensWithCompletion:(void(^)(NSError*, NSDictionary*))]`
+  - Return tokens
+
+        [container getOAuthTokensWithCompletion:^(NSDictionary *tokens){
+          //tokens['com.facebook'] is FB's access token
+        }];
+
+### Platform specific APIs
+
+- `-[SKYContainer loginWithFacebook:(NSDictionary*)options]` and `-[SKYContainer loginWithGoogle:(NSDictionary*)options]`
+  - `options`
+    - uxMode - Either `popup`(default), or `redirect`, popup means in-app-browser (SFSafariViewController/WKWebView) and redirect means Safari.app
+    - clientID
+    - scope
+    - version - FB Client SDK only
+    - cookiePolicy - Google Client SDK only
+    - use3rdPartyClient - whether to use 3rd party client, e.g. Facebook iOS SDK. Default `false`
+  - Returns the user logged in via delegate.
+
+# Changes on API at skygear-server
+
+None.
+
+# Changes on Plugin
+
+## Environment variables
+
+- `UNIQUE_EMAIL_FOR_ACCOUNTS`, boolean
+- `FACEBOOK_CLIENT_SECRET`
+- `GOOGLE_API_KEY`
+- `SSO_ENABLED` e.g `FACEBOOK,GOOGLE`.
+
+Pseudo code of `oauth:handle_access_token`:
+
+```
+function handleAccessToken(token) {
+  // Code should be organized in a way to share logic with auto provider
+  var existingUser = getUserByAccessToken(token);
+  if (existingUser) {
+    return existingUser;
+  }
+  var emailOfThirdPartyAccount = getEmail(token);
+  if (emailOfThirdPartyAccount) {
+    var user = getUserByEmail(emailOfThirdPartyAccount);
+    if (user !== null && UNIQUE_EMAIL_FOR_ACCOUNTS) {
+      return Error("The email is associated with another account");
+    }
+    addAuthToUser(user, token);
+    return user;
+  }
+  return createNewUser(token);
+}
+```
+
+## APIs
+
+- `oauth:auth_url`
+  - Accepts a provider id and options
+  - Return an url for auth
+- `oauth:handle_code`
+  - A handler that accepts code from 3rd party service
+  - Exchange code with access token
+  - Create user if needed
+  - Pass the user back to client
+- `oauth:handle_access_token`
+  - Accepts a provider id, and access token
+  - If this handler is called with a user logged in
+    - Associate the user with auth provider and access token
+  - Otherwise,
+    - Login or create new users according to the provider and access token
+  - Return the user
+
+# Database Scheme
+
+No changes.
```

Comment:
Your comment [here](https://github.com/SkygearIO/features/issues/5#issuecomment-300395577) mentioned:

> I am proposing `oauth:handle_code` is implement at Skygear-server, while other `oauth:handle_access_token` can be model using current `auth:login` with auth provider.

Since we will be using auth provder, I think it make sense to save it in `_user.auth`.